### DISCUSSION TITLE
Negative reconstructed responses with GLM for SCR

### DISCUSSION THREAD

**Discussion by neurok8050** [bachlab/PsPM#440] [Q&A] [answered]
Dear,

fellow PsPM users and developers

I am in the process of learning/using PsPM GLM for SCR for a project. I checked the couple videos that Professor Dominik did and put on the lab website and also looked at the PsPM user manual. I was able to get nice results with the tutorial data, but unfortunately I haven't had luck with other datasets. 

This is the `.m Code` for the batch I used : 
```
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.modelfile = 'test_HRA_1_01_no_filter';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.outdir = {'/home/clauderic/Desktop'};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.chan.chan_def = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.timeunits.markers.mrk_chan = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.datafile = {'/home/clauderic/Maîtrise Psychologie/PsPM-HRA_1 dataset/Data/HRA_1_spike_01.mat'};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.missing.no_epochs = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.data_design.condfile = {'/home/clauderic/Maîtrise Psychologie/PsPM-HRA_1 dataset/Data/onsets_HRA_1_cogent_01.mat'};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.nuisancefile = {''};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.latency.fixed = 'fixed';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.bf.scrf1 = 1;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.norm = false;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.lowpass.disable = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.highpass.disable = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.down = 10;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.direction = 'uni';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.exclude_missing.excl_no = 'No';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.overwrite = false;
```

And this is the result I got for the reconstructed responses (and it is essentially the same think for all of the twenty participants I have) : 
![image](https://user-images.githubusercontent.com/44348995/207406253-6a674341-3b00-4712-acc1-79ca57905f9c.png)

In one of professor Dominik video's, he said that negative reconstructed response are not plausible and they are probably due to bad timing. So I tried to make sure if the onsets and the timing were right. Of what I understand, the `onsets` files are just composed of two vectors the `names` and the `onsets`. The `names` vector contains the names of the different stimuli categories we are interested in. The `onsets` vector contains vectors of the numbers of the time marker from the SCR file. So the markers from the SCR file look like this : 

![image](https://user-images.githubusercontent.com/44348995/207410153-c6cdff82-1a5c-47ed-a675-d5b3ca6b485b.png)
 
So let's say that the 8th, 9th, 13th and the 16th markers from the list above are considered to be aversive stimuli, I would putted them in a vector in the `onsets` file like this : 

![image](https://user-images.githubusercontent.com/44348995/207410990-3147c753-c3a7-4e2d-954f-bdef6f6485a6.png)

The way I find to which stimulus category a marker belongs is by looking at the `cogent` file. In this case I considered CS- to be neutral stimulus and CS+ without US to be aversive (the rest, so CS+ with US is not used). So I go to the `cogent` file the second and the third column correspond respectively to CS (1:CS-; 2:CS+) and US (1:present;0:absent) so essentially any pairs that are (1,0) or (1,1) are CS- and any pairs (2, 0) are CS+ without US. 

![image](https://user-images.githubusercontent.com/44348995/207413383-88ef7bb7-1eee-4ba4-97ea-7dcf89b9e76d.png)

So with the `cogent` file from above I get that the only aversive stimuli (CS+ without US) are the 8th, 9th, 13th and the 16th. The only neutral stimuli are the 1st, 3rd, 4th, 5th, 6th, 7th, 10th, 11th, 12th, 14th and the 17th. The rest is not interesting. That's where the value in the `onsets` example above came from. If I did the right thing to built the `onsets` file, I don't know why I get negative reconstructed responses and I am a bit lost.

Data are from the PsPM-HRA1 open data set and I used the participant number `01` so that would be `HRA_1_spike_01.mat` for the signal and stimuli markers and the `HRA_1_cogent_01.mat` for the information about trials.

I used PsPM version 5.1.0  and the MATLAB version is 9.10.0.1739362 (R2021a) Update 5
 
I know, I didn't use the default filter but that's what I wanted. Even if no filter is apply I am pretty sure that I am not suppose to get negative reconstructed responses. So I would like to understand why I have such negative reconstructed responses or even if it is normal to get them. Sorry if it is a very big text but the devil is in the detail!. Do not hesitate if you want me to provide more information about my problem.

P.S. If I used the default filter setting I get this : 
![image](https://user-images.githubusercontent.com/44348995/207406973-e85ef53c-246d-4b04-8bcf-0755fb474405.png)

Thank you a lot for any help!

Claudéric

**Comment by dominikbach** [comment, marked as answer]
Hi Claudéric

if I understand correctly, you are analysing the HRA1 dataset (https://zenodo.org/record/321641).

In this dataset, each marker corresponds to a CS presentation and there is a US (or US omission) 3.5 s after the CS.

From what I can see, you are only modelling the CS and not the US (or US omission) response. Because the temporal relation of CS and US  is fixed, not random, this is likely to systematically bias your results.

I'd suggest you model both the CS and US (or US omission) responses.

Hope this helps.
Dominik

**Reply by neurok8050** [reply]
Hi Dominik,

I am not sure how to model the CS, US and US omission responses. Do I have to add new marker in the SCR file corresponding to the US based on the CS+ with US (so if I have a CS+ with US at 211.1143, the US would be 214.6143 (CS+ + 3.5 seconds)). Then I add a new vector to the `onsets` vector of the onsets file corresponding to the ith marker of US just like I did for the aversive and neutral marker. Or should I use the DCM even though I want to use the GLM. Or am I totally wrong?

Thank you again for the help, it is deeply appreciated!

Claudéric

**Reply by dominikbach** [reply]
Hi Claudéric

1. For this type of experiment PsPM recommends the non-linear model (DCM) but of course you may want to use GLM for comparison or education purposes.

2. In this particular experiment, not all events have markers, so indeed you'll have to change the setup of the model:

- either change timeunits to "seconds", extract the marker timings to define "CS+, and "CS-" conditions, and add the other events (which, in this experiment, are always 3.5 s after the CS markers) to the US conditions (US, no US)
- or add additional markers to the data file, which might be less transparent

Dominik

**Comment by neurok8050** [comment]
Hi Dominik,

Thank you for the answer I think it worked! 

I ended up using the DCM and following the steps presented in the PsPM manual (essentially creating the timing file like shown in the manual example). I think it did work, I have a model, I don't know if there is a way to confirmed if it worked like with GLM where we can sort of confirmed if it worked or not by looking at the reconstructed responses. 

I started to look at other datasets to practice, I ended up with the SCRV-1 datasets (https://zenodo.org/record/269659). First thing I did was establish if I should use the GLM or the DCM. Not so sure of my answer, but I choose the GLM because there is only a single stimulus presented per trial instead of a CS/US experimental design. Also the example presented in the manual (GLM for SCR tutorial : Appraisal data) have somewhat the same experimental design than the SCRV-1 datasets. 

Yet again, I have negative reconstructed responses this time with the SCRV-1 datasets. My `onsets` file now have 3 `names` (aversive, neutral and startBreak) and 3 `onsets` which are  lists corresponding to the index of trial presenting aversive, neutral pictures or end of block. The startBreak is the index of the last trial before switching block. So trial index corresponding to the presentation of negative picture are put in the `onsets` corresponding to the aversive in `names` . Index of trial presenting neutral picture are put in the `onsets` corresponding to neutral in `names` (just like I did with the HRA-1 datasets earlier). The startBreak correspond to the last trial index before the end of a block, in this case it corresponds to the index 60 and 120 (for participant SCRV_1_cogent_01.mat/SCRV_1_spike_01.mat). I also didn't put the index trial 61 and 121, in the example in the manual the startBreak index are 32 and 64, and the trial index 33 and 65 are not in either the aversive or neutral condition, so I applied the same logic.

This is the `.m Code` I used for the GLM : 

```
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.modelfile = 'test_SCRV_1_no_filter_code_cd_startbreak';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.outdir = {'/home/clauderic/Desktop'};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.chan.chan_def = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.timeunits.markers.mrk_chan = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.datafile = {'/home/clauderic/Maîtrise Psychologie/PsPM-SCRV_1 dataset/Data/SCRV_1_spike_01.mat'};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.missing.no_epochs = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.data_design.condfile = {'/home/clauderic/Maîtrise Psychologie/PsPM-SCRV_1 dataset/Data/onsets_SCRV_1_cogent_01.mat'};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.session.nuisancefile = {''};
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.latency.fixed = 'fixed';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.bf.scrf1 = 1;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.norm = false;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.lowpass.disable = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.highpass.disable = 0;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.down = 10;
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.filter.edit.direction = 'uni';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.exclude_missing.excl_no = 'No';
matlabbatch{1}.pspm{1}.first_level{1}.scr{1}.glm_scr.overwrite = false;
```

This is the reconstructed responses I get : 
![image](https://user-images.githubusercontent.com/44348995/211930176-f6a7cbcd-4f0d-45ea-ac46-6ff5165762bb.png)

If I use the exact same logic but for the trSP1 datasets (the one from the manual example), I get this : 
![image](https://user-images.githubusercontent.com/44348995/211930448-1148f486-7c31-4111-bc66-fe62ddc1085b.png)

Which make me believe that again I might have bad timing or that I should not use GLM with the SCRV-1 datasets.

Again I did not use the default filter setting, and that is what I wanted. Although, if I use the default filter setting I get this : 
![image](https://user-images.githubusercontent.com/44348995/211932422-9662401a-7b09-4096-97d8-92c2519e37ce.png)

Thank you again for all the help it is really appreciated!

Claudéric

**Comment by dominikbach** [comment]
Hi Claudéric 

When using default filter settings all looks fine on your plots. 

Model and response function are build for a specific preprocessing (filtering) pipeline. If you change the filter (in particular, if you remove the highpass) then the impulse response of the system will change substantially - and the default response function will not be appropriate any more - you'd have to build a new one.

Dominik
